### ISSUE TITLE
Accessibility in Complex Forms

### ISSUE THREAD

**Issue by andrewleith** [wet-boew/wet-boew#6535]
I was hoping I could get the community's opinion on a technique we were hoping to use in our complex forms.  Our forms consist of questions and answers.  Sometimes questions can get nested, and sometimes a question will have more than one answer (input/select/etc).

In the past we used the obvious techniques of labels and fieldsets to identify inputs in our forms and this works well in simple situations.  When the legend text gets long (which happens often for us) and the form groups become nested (which also happens often) things do not work so well with screen readers.  There is very inconsistent behavior in how screen readers "read" legends and when they do so.  In practice, there does not seem to be a way to mark up a page with fieldsets that is consistent across browser/screen reader combinations.

To that end we devised what follows, in an effort to provide a consistent, simple approach to screen reader users.  We mark up question text using H3's, and sub-questions as H4s and label each form control using a label.  We also add tab stops on each of the headings, so that tabbing through the form stops and reads every single question, sub-question and supplemental information even in forms mode.  The benefit of this over fieldsets is that it is a) consistent and b) doesnt automatically add verbosity for the screen reader users;  typically, using fieldsets, the entire question is re-read every time they tab to a new field in the group.  Instead, with the approach below, they would simply check what heading they were closest to if they lost their context.

Here is an example of the approach with some longer question text.  If a fieldset was used, the entire text would be read for every form control.

```html
<h1 tabindex="0">Testing Application</h1>
<h2 tabindex="0">Section A: Personal Information</h2>

<h3 class="sc-text" tabindex="0">
    <span class="wb-invisible">Question 1.</span>
    What was the primary address of your business as of January 1, 2014?
    If you have more than one location, use the location that is nearest to Ottawa, Ontario.
</h3>

<label for="STREET_NUMBER">Street number</label>  
<input id="STREET_NUMBER" name="STREET_NUMBER" type="text" />

<label for="STREET_NAME">Street name</label>  
<input id="STREET_NAME" name="STREET_NAME" type="text" />

<label for="STREET_NAME">City</label>  
<input id="CITY" name="CITY" type="text" />
```		

Here is another example with some (conceptual) nesting:

```html
<h1 tabindex="0">Testing Application</h1>
<h2 tabindex="0">Section A: Personal Information</h2>

<h3 class="sc-text" tabindex="0">
    <span class="wb-invisible">Question 1.</span>
    What was the primary address of each of your business locations as of January 1, 2014?
</h3>

<h4>Location 1 Address</h4>
<label for="STREET_NUMBER">Street number</label>  
<input id="STREET_NUMBER" name="STREET_NUMBER" type="text" />

<label for="STREET_NAME">Street name</label>  
<input id="STREET_NAME" name="STREET_NAME" type="text" />

<label for="CITY">City</label>  
<input id="CITY" name="CITY" type="text" />

<h4>Location 2 Address</h4>
<label for="STREET_NUMBER2">Street number</label>  
<input id="STREET_NUMBER2" name="STREET_NUMBER2" type="text" />

<label for="STREET_NAME2">Street name</label>  
<input id="STREET_NAME2" name="STREET_NAME2" type="text" />

<label for="CITY2">City</label>  
<input id="CITY2" name="CITY2" type="text" />

```	

Please reply if you have any comments or suggestions.

Thanks,
Andrew

**Comment by LaurentGoderre** [comment]
The issue with this particular example is that you have multiple field with the same label and only a header to tell them apart. A fieldset is the safest element to use here (and you can style it too look the same as if you were using headers)

**Comment by andrewleith** [comment]
Thanks for the feedback.  You didn't catch what we're trying to do, however: make the experience for a screen reader user less verbose and more natural.  Adding the fieldsets results in some painful usability and unnecessary verbosity for screen reader users.

You're right - if a user were to arbitrarily land on an input, they would not "automatically" know that the CITY input had to do with Location 1 or Location 2, they would have to read the heading to get it.  The idea though is that they fill out the form the top down, so context IS given as they navigate, verbosity is eliminated, and context can be obtained again later through headings if needed.

While the fieldset is the "safest", I still feel its not all that good.

**Comment by LaurentGoderre** [comment]
@andrewleith in forms mode in some screen reader non form elements get ignored so it's possible that the user might not get the information at all. Besides, there are scenario where filling the form is not sequential, for example when fixing an submission error.

We should be weary of making judgments about usability of screen readers if we are not everyday users. Verbosity issues should be addressed at the screen reader level (like is done for punctuation, the user can select the level of information that is announced)

**Comment by andrewleith** [comment]
In our testing (JAWS + NVDA, IE + Firefox)  the information was presented the same way as if you read down the page - nothing was omitted.

In the case of submission errors we can easily annotate the fields with the missing context if that was necessary.

In the end though I think you (like most developers) want to shy away from trying to improve the screenreader experience because of how fragmented and inconsistent the screen readers themselves are, and I can understand that.

I think I will sprinkle in some aria-describedby's and call it a day.  Do you think that is sufficient for context, instead of using fieldsets?

**Comment by LaurentGoderre** [comment]
I don't want to shy away from improving the experience, I am only saying that I can't judge the quality of the experience or what would make it better or worse because I am not a regular user of those technology.

Maybe @dfarough or @mysticjeff  has some insight on this matter.

**Comment by andrewleith** [comment]
Thanks, here is the first example in a fiddle if it makes it any easier:

https://jsfiddle.net/6e9ozv5o/embedded/result/

**Comment by mysticjeff** [comment]
We did just have this discussion at AWG and I believe that the consensus  was that describedby or an alert (like what we do for error messages in WET) would be preferable.


-----Original Message-----

**Comment by mysticjeff** [comment]
I'll have a look tomorrow in case I'm totally off 

-----Original Message-----

**Comment by andrewleith** [comment]
@mysticjeff Thanks Jeff.  I am working on two links on jsbin so you can experience the difference between the two techniques.  I'll let you know as soon as they're ready.

**Comment by mysticjeff** [comment]
That's perfect; I can run a few people through it and give some feedback 

-----Original Message-----

**Comment by andrewleith** [comment]
@mysticjeff: here are the links:

1. As described above, no `fieldset`, no `aria-describedby`: http://jsbin.com/cinigogizi
2. With `aria-describedby` from the inputs to the question: http://jsbin.com/gumujifetu

Thanks for taking the time to review this.

Andrew

**Comment by dfarough** [comment]
These examples are interesting.
For the most part, from my prospective as an advanced Jaws user, the first example minus the ARIA is more than adequate for my purposes.  Of course the context is important and the more complex the form the more important it is to insure that the user gets all the necessary information when moving through the form.

I like example1 for the following reasons.

1.  Tabbing through the headings is intuitive.  Even less experienced users will benefit  from the tabular access to headings within the question.
2.  focus is obvious  for low vision users.

For example 2 with ARIA I see both Pros and Cons.
PRO: Question is provided when tabbing to each field.
 Con: a little too verbose for my liking but might be appropriate depending on the level of complication in the  form. design. I noticed another thing however  that works for me when using a form that is designed this way.  If you use the arrows instead of tab to move from one field to another you do not hear the question each time.  If you use the field list in Jaws to select the field you want, you will get to the field  directly without hearing the question or label.  so this approach does allow for the most flexibility for users.

**Comment by andrewleith** [comment]
Thanks @dfarough. So is your conclusion that, to be safe, we should maintain the `aria-describedby` even at the cost of the verbosity to make sure less advanced users can recall the context?

**Comment by dfarough** [comment]
If the context warrants then this method will insure that no information is missed.  Important instructions such as those found when creating passwords are good candidates for this approach.  Given the example you presented, I feel that method 1 with the focusable headings  is sufficient.

**Comment by LaurentGoderre** [comment]
Can we close then?

**Comment by andrewleith** [comment]
I was hoping we might get some input from other community members.  @mysticjeff: does anyone have anything else to add? or could we close this?

**Comment by mysticjeff** [comment]
I have 2 other people testing this; but they won't get to it until thurs/fri (likely)


-----Original Message-----

**Comment by LaurentGoderre** [comment]
ok, np

**Comment by mysticjeff** [comment]
the links provided are blocked by the network: http://jsbin.com/cinigogizi, http://jsbin.com/gumujifetu.

Could you zip it up and dropbox/googledrive/onedrive it and we'll set it up here 

-----Original Message-----

**Comment by andrewleith** [comment]
Our department conveniently blocks all three of those services, i'll see what I can do from my phone :)

**Comment by andrewleith** [comment]
Can you download it from github and then set it up there?

1. no_fieldset_no_aria_describedby: https://gist.github.com/andrewleith/10086c914a6c33c22833/download
2.  with_aria_describedby: https://gist.github.com/andrewleith/1373e36cb7ac2ff1c124/download

**Comment by mysticjeff** [comment]
Today, we did a number of user tests with different people, different technologies and different skill levels, of the form with and without the use of ARIA and we agree with @dfarough, who said that: " a little too verbose for my liking but might be appropriate depending on the level of complication in the form. design.".

We would further emphasize the negative experience this level of verbosity & repetition introduces to the user.
repeating the exact text in a form control in  the  “aria-describedby” where a description is intended to provide more verbose information would be a little overkill and will cause an annoyance especially on short forms. 

Where adding “aria-describedby” would be beneficial is long forms especially the one that use sections. But only if it is not a duplication of information that is provided through another mechanism and deemed "essential".    Additionally, it should be provided in a reverse hierarchy (i.e. the most important information first). If there is a duplication of the same information, which we don't recommend, have a mechanism to advise the user of this duplication.  Additionally, you might explore a method of toggling on/off the ARIA markup on demand (or choose this level of verbosity in advance).




-----Original Message-----

**Comment by hsrudnicki** [comment]
Are you ready to close this issue or otherwise?

**Comment by andrewleith** [comment]
Yes, thank you all for your feedback.
